Improve scalability of identity allocation

tgraf opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4560)
<!-- Reviewable:end -->

rlenglet commented on pkg/identity/cache.go:
```diff
@@ -63,13 +65,44 @@ func GetIdentities() []*models.Identity {
+
+			// keep critical section as small as possible
+			needPolicyUpdateMutex.Lock()
+			triggerPolicyUpdate = needPolicyUpdate
```
Remove the variable definition above and just do this here:
```
triggerPolicyUpdate := needPolicyUpdate
```

rlenglet commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -279,6 +288,16 @@ func NewAllocator(basePath string, typ AllocatorKey, opts ...AllocatorOption) (*
+// read while NewAllocator() is being called to ensure that the channel does
+// not block indefinitely while NewAllocator() emits events on it while
+// populating the initial cache.
+func WithEvents(events AllocatorEventChan) AllocatorOption {
```
Mention this new option in the comment for `NewAllocator`.

tgraf replied:
Already there:
```
// The allocator can be configured by passing in additional options:
//  - WithEvents() - enable Events channel
[...]
```

It's confusing because we once had a WithEvents() and it was removed without removing the comment.

rlenglet replied:
👍

aanm commented on pkg/identity/cache.go:
```diff
@@ -63,13 +65,44 @@ func GetIdentities() []*models.Identity {
-func identityWatcher(owner IdentityAllocatorOwner) {
+func identityWatcher(owner IdentityAllocatorOwner, events allocator.AllocatorEventChan) {
+	var (
+		needPolicyUpdateMutex lock.RWMutex
```
`s/RWMutex/Mutex` (RWMutex is slower if you don't do any reads. Yay more ops/s :tada: )

aanm commented on pkg/identity/cache.go:
```diff
@@ -63,13 +65,44 @@ func GetIdentities() []*models.Identity {
 		case kvstore.EventTypeCreate, kvstore.EventTypeDelete:
-			owner.TriggerPolicyUpdates(true)
+			needPolicyUpdateMutex.Lock()
+			needPolicyUpdate = true
```
Wait, can't this be a channel? If this becomes a channel, you don't need to to a time.Sleep on line 91

tgraf replied:
The major change here is to avoid triggering a policy for every event so aggregation has to happen. We can do a unbuffered channel with non-blocking writes though.

aanm commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -778,16 +802,27 @@ func (a *Allocator) startWatch() waitChan {
 						kvstore.Trace("Removing id from cache", nil, logrus.Fields{fieldID: id})
 						delete(a.nextCache, id)
+
+						// key value is not available. Need to scan the cache
```
It's not available because you have delete it in `808`.

Can't you do:
```
key := a.nextCache[id]
delete(a.nextKeyCache, key)
delete(a.nextCache, id)
```

tgraf replied:
Good point. Will do.

aanm commented on pkg/kvstore/allocator/allocator_test.go:
```diff
@@ -100,7 +100,7 @@ func (s *AllocatorSuite) TestSelectID(c *C) {
 func (s *AllocatorSuite) BenchmarkAllocate(c *C) {
 	allocatorName := randomTestName()
-	maxID := ID(c.N)
+	maxID := ID(256 + c.N)
```
There's a `identity.MinimalNumericIdentity.Uint32()`

tgraf replied:
We can't import it without a massive import cycle refactoring. We have to rethink our packages structure.

tgraf committed:
test: Use latest stable etcd and consul images

* etcd 3.2.17
* consul 1.1.0

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
identity: Process identity events in batches

The existing watcher in charge of detecting identity changes in the cluster was
implementing naive behavior. It was simply waiting while reading on a go
channel to process create and modify events for new or changed identities. It
would then call the function to trigger recalculation of required policies. The
function would block until the operation completed. It would then continue
reading and process whatever events had accumulated in the meantime.

This is suboptimal for various reasons:

* Events can pile up quicker than we can process as policy recalculation is an
  expensive operation. This would eventually cause the allocator code to slow
  down causing lag in handling policy update triggers on remote nodes.

* Causing unnecessary load as a policy trigger resolves the necessary work for
  all changed or new identities in the cache since the function was last
  called.

Optimize this by making the channel handler lightweight by using the new
trigger package. The update continues to be blocking as there is no point in
running parallel policy updates. Policy updates are triggered at most once per
second as this is an expensive operation and batching helps reduce overall load
which will reduce build times and thus latency.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
allocator: Avoid scanning sequentual list when allocating

Testing with a kvstore with 25K identities pre-loaded

Before:
allocations time-span 1.057793212s success 36.869210/s error 0.000000/s

After:
allocations time-span 1.00054621s success 1009.448629/s error 0.000000/s

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
allocator: benchmark: Reserve ID space for reserved identities

Depending on the c.N automatically chosen by the test framework, the MaxID was
able to drop below the MinID which caused the test to fail sometimes.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
trigger: New trigger package

Simple logic to implement trigger based invocation of background routines that
need to be serialized and should be subject to a minimum interval between
invocations.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
identity: Fix allocator init with more than pre-existing 1024 keys

The current allocator init code issues a ListAndWatch operation against the
kvstore to populate the local cache of identities. It also emits an event on
a channel for each cache entry to indicate the event of a new identity.

So far, this channel was created by the allocator and returned. The buffer size
was 1024. Since the channel was never read before the allocator was returned to
the caller, the buffer size acted as an upper boundary for number of identities
that can exist in the kvstore while an allocator can initialize.

Fix this by making the caller responsible for creating the events channel and
have the caller pass it in with a WithEvents() function.

Add a section in the function documentation to mention the requirement to read
the event channel while initializing is happening.

Fixes: #4557

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm commented on pkg/trigger/trigger.go:
```diff
@@ -0,0 +1,134 @@
+			}
+
+			if t.params.TriggerFunc != nil {
+				t.params.TriggerFunc()
```
I was about to leave a comment, decided not to, but then I read your comment:

>@aanm Sorry about that, I've overlooked the initial two comments. I've changed the logic to provide a wakeup channel but added a minimum interval in between policy updates of 1 second to avoid continuous policy update cycles in environments with lots of identity changes to leave CPU for actual builds.


My initial thought:
if `t.params.TriggerFunc()` takes more time than `t.params.MinInterval`, the `t.params.TriggerFunc()` will always be executed for continuous triggers. Is this the expected behavior?

tgraf replied:
Yes
